File: src/R/systems/validate_phase7b_freeze.R
#!/usr/bin/env Rscript
# Independent base-R freeze-boundary validation for accepted Phase 7B.
args <- commandArgs(trailingOnly=TRUE)
root <- if (length(args)) normalizePath(args[[1]], mustWork=TRUE) else normalizePath(".", mustWork=TRUE)
reports <- file.path(root, "reports")
source_commit <- "39d9c97ebab0dad736188a320dc037a602e5f594"
final_name <- "phase7b_exposure_dependencies_controls_freeze_manifest.json"
sibling_name <- "phase7c_environmental_health_futures_freeze_manifest.json"
working_name <- "exposure_dependency_manifest.json"
check_name <- "exposure_dependency_artifact_check.json"
strict_name <- "phase7b_strict_provenance_check.json"
phase7a_name <- "phase7a_exposure_environmental_health_freeze_manifest.json"
review_initial <- "phase7b_exposure_dependencies_independent_review_initial.md"
review_final <- "phase7b_exposure_dependencies_independent_review.md"
text_ext <- c("csv", "json", "md", "txt", "yml", "yaml", "svg", "py", "r", "R", "toml")
expected_package <- c(
  "data/processed/networks/exposure_dependency_edges.csv",
  "data/processed/analysis/exposure_control_register.csv",
  "data/processed/analysis/exposure_evidence_strength_register.csv",
  "data/processed/analysis/exposure_dependency_control_matrix.csv",
  "outputs/maps/systems/24_exposure_dependencies_controls_2026.png",
  "outputs/maps/systems/24_exposure_dependencies_controls_2026.svg",
  "reports/exposure_dependency_sources.md",
  "reports/exposure_dependency_assumptions.md",
  "reports/exposure_dependency_findings.md",
  "reports/exposure_evidence_limits.md",
  "reports/exposure_dependency_qa.md",
  "reports/phase7b_citation_ledger.json",
  "reports/phase7b_strict_provenance.md",
  "reports/phase7b_strict_provenance_check.json",
  "src/python/systems/build_exposure_dependencies.py",
  "src/python/systems/validate_exposure_dependencies.py",
  "src/R/systems/validate_exposure_dependencies.R",
  "reports/exposure_dependency_manifest.json",
  "reports/exposure_dependency_artifact_check.json",
  "reports/phase7b_exposure_dependencies_independent_review_initial.md",
  "reports/phase7b_exposure_dependencies_independent_review.md",
  "docs/phase_briefs/phase7b_exposure_dependencies_controls.md",
  "src/python/systems/validate_phase7b_freeze.py",
  "src/R/systems/validate_phase7b_freeze.R"
)
expected_prior_entries <- 435L
expected_prior_unique <- 426L
stop_if <- function(condition, message) if (!isTRUE(condition)) stop(message, call.=FALSE)
json_text <- function(path) paste(readLines(path, warn=FALSE, encoding="UTF-8"), collapse=" ")
raw_bytes <- function(path) readBin(path, "raw", n=as.numeric(file.info(path)$size))
canonical_raw <- function(raw) {
  vals <- as.integer(raw); next_vals <- c(vals[-1L], NA_integer_); keep <- !(vals == 13L & next_vals == 10L); out <- vals[keep]; out[out == 13L] <- 10L; as.raw(out)
}
crlf_raw <- function(raw) {
  vals <- as.integer(raw); out <- unlist(lapply(vals, function(x) if (x == 10L) c(13L,10L) else x), use.names=FALSE); as.raw(out)
}
sha256_many <- function(paths) {
  paths <- unique(as.character(paths)); if (!length(paths)) return(setNames(character(), character()))
  cmd <- Sys.which("sha256sum")
  if (!nzchar(cmd)) {
    certutil <- Sys.which("certutil"); stop_if(nzchar(certutil), "Neither sha256sum nor certutil is available")
    ans <- vapply(paths, function(path) {
      out <- system2(certutil, c("-hashfile", path, "SHA256"), stdout=TRUE, stderr=TRUE)
      hits <- regmatches(paste(out, collapse=" "), gregexpr("[0-9a-fA-F]{64}", paste(out, collapse=" "), perl=TRUE))[[1]]
      stop_if(length(hits) >= 1L, paste("certutil failed", path)); tolower(hits[[1]])
    }, character(1)); return(setNames(ans, paths))
  }
  ans <- character(); starts <- seq(1L, length(paths), by=40L)
  for (start in starts) {
    chunk <- paths[start:min(start+39L, length(paths))]
    out <- system2(cmd, chunk, stdout=TRUE, stderr=TRUE)
    text <- paste(out, collapse="\n")
    hits <- regmatches(text, gregexpr("[0-9a-fA-F]{64}", text, perl=TRUE))[[1]]
    stop_if(length(hits) == length(chunk), paste("sha256sum failed", start))
    ans <- c(ans, setNames(tolower(hits), chunk))
  }
  ans
}
manifest_entries <- function(path) {
  lines <- readLines(path, warn=FALSE, encoding="UTF-8")
  collection <- if (any(grepl('"artifacts"[[:space:]]*:[[:space:]]*\\{', lines, perl=TRUE))) "artifacts" else "files"
  start <- which(grepl(paste0('"', collection, '"'), lines, fixed=TRUE) & grepl("{", lines, fixed=TRUE))[1]
  stop_if(!is.na(start), paste("missing manifest collection", path))
  rel <- character(); hashes <- character(); bytes <- numeric(); current <- NA_character_; current_hash <- NA_character_; current_bytes <- NA_real_
  flush <- function() {
    if (!is.na(current) && !is.na(current_hash)) { rel <<- c(rel, current); hashes <<- c(hashes, current_hash); bytes <<- c(bytes, current_bytes) }
    current <<- NA_character_; current_hash <<- NA_character_; current_bytes <<- NA_real_
  }
  if (start + 1L <= length(lines)) for (line in lines[(start + 1L):length(lines)]) {
    if (grepl("^  \\},?[[:space:]]*$", line)) { flush(); break }
    direct <- regexec('^[[:space:]]*"([^"]+/[^"]+)"[[:space:]]*:[[:space:]]*"([0-9a-f]{64})"', line, perl=TRUE); hit <- regmatches(line, direct)[[1]]
    if (length(hit) == 3L) { flush(); rel <- c(rel, hit[[2]]); hashes <- c(hashes, hit[[3]]); bytes <- c(bytes, NA_real_); next }
    nested <- regexec('^[[:space:]]*"([^"]+/[^"]+)"[[:space:]]*:[[:space:]]*\\{', line, perl=TRUE); hit <- regmatches(line, nested)[[1]]
    if (length(hit) == 2L) { flush(); current <- hit[[2]] }
    h <- regexec('"sha256"[[:space:]]*:[[:space:]]*"([0-9a-f]{64})"', line, perl=TRUE); hh <- regmatches(line, h)[[1]]
    if (length(hh) == 2L && !is.na(current)) current_hash <- hh[[2]]
    b <- regexec('"bytes"[[:space:]]*:[[:space:]]*([0-9]+)', line, perl=TRUE); bb <- regmatches(line, b)[[1]]
    if (length(bb) == 2L && !is.na(current)) current_bytes <- as.numeric(bb[[2]])
  }
  flush(); data.frame(rel=rel, sha256=hashes, bytes=bytes, stringsAsFactors=FALSE)
}
verify_entry_sets <- function(sets) {
  all_rows <- do.call(rbind, lapply(sets, function(x) x$entries))
  stop_if(nrow(all_rows) > 0L, "no manifest entries")
  paths <- file.path(root, all_rows$rel)
  raw_hashes <- sha256_many(paths)
  normalized <- list(); crlf <- list(); need <- character()
  for (i in seq_len(nrow(all_rows))) {
    path <- paths[[i]]; expected <- all_rows$sha256[[i]]; size <- all_rows$bytes[[i]]
    raw <- raw_bytes(path); raw_hash <- unname(raw_hashes[[path]])
    raw_ok <- identical(raw_hash, expected) && (is.na(size) || as.numeric(size) == length(raw))
    if (!raw_ok && tolower(tools::file_ext(path)) %in% tolower(text_ext)) {
      need <- c(need, path)
      if (is.null(normalized[[path]])) {
        lf <- canonical_raw(raw); cf <- crlf_raw(lf)
        p1 <- tempfile(fileext=".txt"); p2 <- tempfile(fileext=".txt"); writeBin(lf, p1); writeBin(cf, p2)
        normalized[[path]] <- c(path=p1, bytes=length(lf)); crlf[[path]] <- c(path=p2, bytes=length(cf))
      }
    } else stop_if(raw_ok, paste("artifact hash/bytes", all_rows$rel[[i]]))
  }
  if (length(need)) {
    lf_paths <- vapply(normalized[unique(need)], function(x) unname(x[["path"]]), character(1)); cf_paths <- vapply(crlf[unique(need)], function(x) unname(x[["path"]]), character(1))
    lf_hash <- sha256_many(lf_paths); cf_hash <- sha256_many(cf_paths)
    for (i in seq_len(nrow(all_rows))) {
      path <- paths[[i]]; expected <- all_rows$sha256[[i]]; size <- all_rows$bytes[[i]]
      if (path %in% unique(need)) {
        l <- normalized[[path]]; c <- crlf[[path]]; l_path <- unname(l[["path"]]); c_path <- unname(c[["path"]])
        ok <- (unname(lf_hash[[l_path]]) == expected && (is.na(size) || as.numeric(size) == as.numeric(l[["bytes"]]))) || (unname(cf_hash[[c_path]]) == expected && (is.na(size) || as.numeric(size) == as.numeric(c[["bytes"]])))
        stop_if(ok, paste("text artifact hash/bytes", all_rows$rel[[i]]))
      }
    }
  }
  invisible(TRUE)
}
read_csv <- function(relative) read.csv(file.path(root, relative), stringsAsFactors=FALSE, check.names=FALSE, na.strings=c(""))
check_png <- function(relative) { x <- readBin(file.path(root, relative), "raw", n=8); stop_if(identical(as.integer(x), c(137L,80L,78L,71L,13L,10L,26L,10L)), relative); stop_if(file.info(file.path(root, relative))$size > 100, relative) }
check_svg <- function(relative, terms) { txt <- paste(readLines(file.path(root, relative), warn=FALSE, encoding="UTF-8"), collapse=" "); for (term in terms) stop_if(grepl(term, txt, fixed=TRUE), paste(relative, term)) }
phase7a <- manifest_entries(file.path(reports, phase7a_name)); stop_if(nrow(phase7a) == 8L, "Phase 7A manifest artifact count")
phase7a_text <- json_text(file.path(reports, phase7a_name)); for (term in c('"accepted_phase": "7A"','"status": "ACCEPTED / FROZEN"','"nodes": 20','"edges": 20','"pathway_register": 5','"monitoring_matrix_rows": 5','"evidence_crosswalk": 10','"uncertainty_register": 13')) stop_if(grepl(term, phase7a_text, fixed=TRUE), paste("Phase 7A", term))
final_text <- json_text(file.path(reports, final_name)); for (term in c('"accepted_phase": "7B"','"baseline": "Exposure Dependencies, Evidence Strength & Controls, 2026"','"status": "ACCEPTED / FROZEN"','"accepted_date": "2026-09-10"',paste0('"source_commit": "',source_commit,'"'), '"dependency_edges": 20','"controls": 7','"evidence_rows": 5','"phase7a_immutable": true','"phase14_implemented": false','"great_black_swamp": "C — HOLD / noncanonical"','"toledo_intake_coordinate_discrepancy": "UNRESOLVED"')) stop_if(grepl(term, final_text, fixed=TRUE), paste("final manifest", term))
stop_if(grepl('"matrix_semantic_shape":', final_text, fixed=TRUE) && grepl('5', final_text, fixed=TRUE) && grepl('10', final_text, fixed=TRUE), "final matrix semantic shape")
stop_if(grepl('"matrix_stored_shape":', final_text, fixed=TRUE) && grepl('5', final_text, fixed=TRUE) && grepl('11', final_text, fixed=TRUE), "final matrix stored shape")
final <- manifest_entries(file.path(reports, final_name)); stop_if(identical(as.character(final$rel), expected_package), "Phase 7B final artifact inventory"); stop_if(any(final$rel == "src/python/systems/validate_phase7b_freeze.py"), "Python freeze validator missing")
working <- manifest_entries(file.path(reports, working_name)); stop_if(identical(as.character(working$rel), expected_package[1:6]), "Phase 7B working artifact inventory")
working_text <- json_text(file.path(reports, working_name)); for (term in c('"phase": "7B"','"status": "validated_working_package"','"dependency_edges": 20','"control_register": 7','"evidence_register": 5','"matrix": [','5,','10')) stop_if(grepl(term, working_text, fixed=TRUE), paste("working manifest", term))
check_text <- json_text(file.path(reports, check_name)); for (term in c('"status": "passed"','"dependency_edges": 20','"control_register": 7','"evidence_register": 5','"matrix_dimensions": [ 5, 11 ]','"map24_valid": true','"phase7a_immutable": true','"no_exposure_score": true','"no_dose_or_health_model": true')) { if (term == '"matrix_dimensions": [ 5, 11 ]') stop_if(grepl('"matrix_dimensions": [', check_text, fixed=TRUE) && grepl('5,', check_text, fixed=TRUE) && grepl('11', check_text, fixed=TRUE), "artifact matrix") else stop_if(grepl(term, check_text, fixed=TRUE), paste("artifact check", term)) }
strict_text <- json_text(file.path(reports, strict_name)); for (term in c('"status": "passed"','"phase": "7B"','"working_manifest_hashes_match": true','"unsupported_exact_health_outcome_claims": false','"unsupported_exposure_dose_illness_claims": false')) stop_if(grepl(term, strict_text, fixed=TRUE), paste("strict check", term))
sets <- list(list(name="final", entries=final), list(name="working", entries=working), list(name="Phase 7A", entries=phase7a))
phase7b_manifest <- jsonlite_missing <- NULL
phase_files <- list.files(reports, pattern="^phase.*_freeze_manifest[.]json$", full.names=TRUE); phase_files <- phase_files[!basename(phase_files) %in% c(final_name, sibling_name)]
prior <- lapply(phase_files, function(path) { x <- manifest_entries(path); sets[[length(sets)+1L]] <<- list(name=basename(path), entries=x); x })
stop_if(length(phase_files) == 29L, "prior manifest count")
all_prior_rel <- unique(unlist(lapply(prior, function(x) x$rel)))
stop_if(sum(vapply(prior, nrow, integer(1))) == expected_prior_entries, "prior manifest entries")
stop_if(length(all_prior_rel) == expected_prior_unique, "prior unique artifacts")
verify_entry_sets(sets)
changed <- gsub("\\\\", "/", system2("git", c("-C", root, "diff", "--name-only", source_commit), stdout=TRUE, stderr=TRUE)); stop_if(length(intersect(changed, all_prior_rel)) == 0L, "protected prior artifact changed")
check_png("outputs/maps/systems/24_exposure_dependencies_controls_2026.png"); check_svg("outputs/maps/systems/24_exposure_dependencies_controls_2026.svg", c("MAP 24", "Environmental condition", "Evidence gap"))
d <- read_csv("data/processed/networks/exposure_dependency_edges.csv"); controls <- read_csv("data/processed/analysis/exposure_control_register.csv"); e <- read_csv("data/processed/analysis/exposure_evidence_strength_register.csv"); m <- read_csv("data/processed/analysis/exposure_dependency_control_matrix.csv"); n <- read_csv("data/processed/networks/exposure_context_nodes.csv")
stop_if(nrow(d)==20L && nrow(controls)==7L && nrow(e)==5L && nrow(m)==5L && ncol(m)==11L, "Phase 7B counts")
stop_if(setequal(unique(d$pathway_family), c('Drinking Water / HAB','Ambient Air','Soil / Groundwater / Legacy Contamination','Food / Fish / Recreational Water','Heat')), "Phase 7B pathway families")
stop_if(all(e$exposure_confirmation=='unconfirmed') && all(e$dose_information=='unknown') && all(e$health_outcome_information=='unknown'), "Phase 7B health boundary")
initial <- json_text(file.path(reports, review_initial)); final_review_text <- json_text(file.path(reports, review_final)); for (term in c('"passed": false')) stop_if(grepl(term, initial, fixed=TRUE), "initial review preserved"); for (term in c('"passed": true','"security_concerns": []','"logic_errors": []','"provenance_errors": []','"exposure_boundary_errors": []','"dependency_errors": []','"spatial_scale_errors": []','"schema_errors": []')) stop_if(grepl(term, final_review_text, fixed=TRUE), paste("final review", term))
status_terms <- list(
  "PROJECT_STATUS.md"=c("## phase 7b — accepted / frozen","## phase 7c — accepted / frozen",final_name,sibling_name,"active phase: **none**","phase 14 is not implemented","next approved planning target","great black swamp","hold","noncanonical","intake-coordinate discrepancy","unresolved"),
  "docs/canon_status.md"=c("phase 7b is **accepted / frozen**","phase 7c is **accepted / frozen**",final_name,sibling_name,"active phase: **none**","phase 14 is not implemented","next approved planning target","great black swamp","hold","noncanonical","intake-coordinate discrepancy","unresolved"),
  "reports/current_phase_handoff.md"=c("## phase 7b/7c final acceptance / freeze handoff","phase 7b: **accepted / frozen**","phase 7c: **accepted / frozen**",final_name,sibling_name,"active phase: **none**","phase 14 is not implemented","next approved planning target","great black swamp","hold","noncanonical","intake-coordinate discrepancy","unresolved"),
  "README.md"=c("phase 7b is **accepted / frozen**","phase 7c is **accepted / frozen**",final_name,sibling_name,"phase 14 is not implemented"),
  "reports/README.md"=c("phase 7b is **accepted / frozen**","phase 7c is **accepted / frozen**",final_name,sibling_name),
  "CHANGELOG.md"=c("sol formally accepted and froze phase 7b","sol formally accepted and froze phase 7c",final_name,sibling_name,"active phase is none","phase 14 is not implemented"),
  "docs/phase_briefs/phase7b_exposure_dependencies_controls.md"=c("status: accepted / frozen","phase7b_exposure_dependencies_controls_freeze_manifest.json"),
  "docs/phase_briefs/phase7c_environmental_health_futures.md"=c("status: accepted / frozen",sibling_name)
)
for (relative in names(status_terms)) { txt <- tolower(json_text(file.path(root, relative))); for (term in status_terms[[relative]]) stop_if(grepl(tolower(term), txt, fixed=TRUE), paste("status", relative, term)) }
for (base in c(file.path(root,"data","processed"),file.path(root,"outputs","maps","systems"),file.path(root,"src","python","systems"),file.path(root,"src","R","systems"))) { files <- list.files(base, recursive=TRUE, full.names=TRUE); bad <- files[grepl("phase14|phase_14",tolower(basename(files))) | grepl("(^|[/\\\\])44_",tolower(basename(files)))]; stop_if(length(bad)==0L, paste("Phase 14 implementation", base)) }
cat(sprintf("Phase 7B R freeze validation passed: 24 final artifacts; 6 working artifacts; Phase 7A artifacts %d; prior freeze inventory %d/%d; review, status surfaces, holds, deferred maintenance, and no Phase 14 implementation verified\n", nrow(phase7a), expected_prior_entries, expected_prior_unique))

File: src/python/systems/validate_phase7b_freeze.py
#!/usr/bin/env python3
"""Independent freeze-boundary validation for accepted Phase 7B."""
from __future__ import annotations

import argparse
import csv
import hashlib
import json
import re
import struct
import subprocess
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Any

SOURCE_COMMIT = "39d9c97ebab0dad736188a320dc037a602e5f594"
FINAL_MANIFEST = "reports/phase7b_exposure_dependencies_controls_freeze_manifest.json"
SIBLING_FINAL_MANIFEST = "reports/phase7c_environmental_health_futures_freeze_manifest.json"
WORKING_MANIFEST = "reports/exposure_dependency_manifest.json"
ARTIFACT_CHECK = "reports/exposure_dependency_artifact_check.json"
STRICT_CHECK = "reports/phase7b_strict_provenance_check.json"
PHASE7A_MANIFEST = "reports/phase7a_exposure_environmental_health_freeze_manifest.json"
REVIEW_INITIAL = "reports/phase7b_exposure_dependencies_independent_review_initial.md"
REVIEW_FINAL = "reports/phase7b_exposure_dependencies_independent_review.md"
TEXT_SUFFIXES = {".csv", ".json", ".md", ".py", ".r", ".txt", ".yaml", ".yml", ".toml", ".svg"}
FAMILIES = {
    "Drinking Water / HAB",
    "Ambient Air",
    "Soil / Groundwater / Legacy Contamination",
    "Food / Fish / Recreational Water",
    "Heat",
}
EXPECTED_PRIOR_ENTRIES = 435
EXPECTED_PRIOR_UNIQUE = 426
EXPECTED_PACKAGE_ARTIFACTS = (
    "data/processed/networks/exposure_dependency_edges.csv",
    "data/processed/analysis/exposure_control_register.csv",
    "data/processed/analysis/exposure_evidence_strength_register.csv",
    "data/processed/analysis/exposure_dependency_control_matrix.csv",
    "outputs/maps/systems/24_exposure_dependencies_controls_2026.png",
    "outputs/maps/systems/24_exposure_dependencies_controls_2026.svg",
    "reports/exposure_dependency_sources.md",
    "reports/exposure_dependency_assumptions.md",
    "reports/exposure_dependency_findings.md",
    "reports/exposure_evidence_limits.md",
    "reports/exposure_dependency_qa.md",
    "reports/phase7b_citation_ledger.json",
    "reports/phase7b_strict_provenance.md",
    STRICT_CHECK,
    "src/python/systems/build_exposure_dependencies.py",
    "src/python/systems/validate_exposure_dependencies.py",
    "src/R/systems/validate_exposure_dependencies.R",
    WORKING_MANIFEST,
    ARTIFACT_CHECK,
    REVIEW_INITIAL,
    REVIEW_FINAL,
    "docs/phase_briefs/phase7b_exposure_dependencies_controls.md",
    "src/python/systems/validate_phase7b_freeze.py",
    "src/R/systems/validate_phase7b_freeze.R",
)


def canonical_text(data: bytes) -> bytes:
    return data.replace(b"\r\n", b"\n").replace(b"\r", b"\n")


def sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def artifact_entries(payload: dict[str, Any]) -> dict[str, Any]:
    return payload.get("artifacts", payload.get("files", {}))


def load_json(root: Path, relative: str) -> dict[str, Any]:
    with (root / relative).open(encoding="utf-8") as handle:
        return json.load(handle)


def verify_hash(root: Path, relative: str, expected: dict[str, Any] | str) -> str:
    path = root / relative
    assert path.is_file() and path.stat().st_size > 0, f"missing artifact: {relative}"
    expected_hash = expected if isinstance(expected, str) else str(expected["sha256"])
    expected_bytes = None if isinstance(expected, str) else expected.get("bytes")
    raw = path.read_bytes()
    candidates = [(sha256(raw), len(raw), "exact")]
    if path.suffix.lower() in TEXT_SUFFIXES:
        normalized = canonical_text(raw)
        crlf = normalized.replace(b"\n", b"\r\n")
        candidates.extend(
            [
                (sha256(normalized), len(normalized), "newline-normalized"),
                (sha256(crlf), len(crlf), "newline-normalized"),
            ]
        )
    for digest, size, mode in candidates:
        if digest == expected_hash and (expected_bytes is None or int(expected_bytes) == size):
            return mode
    raise AssertionError(f"hash/byte mismatch: {relative}")


def verify_artifacts(root: Path, payload: dict[str, Any], label: str, expected: tuple[str, ...] | None = None) -> dict[str, int]:
    artifacts = artifact_entries(payload)
    assert artifacts, f"{label}: no artifacts"
    if expected is not None:
        assert tuple(artifacts) == expected, f"{label}: artifact inventory"
    exact = 0
    normalized = 0
    for relative, metadata in artifacts.items():
        mode = verify_hash(root, relative, metadata)
        if mode == "exact":
            exact += 1
        else:
            normalized += 1
    return {"entries": len(artifacts), "exact": exact, "newline_normalized": normalized}


def read_csv(root: Path, relative: str) -> list[dict[str, str]]:
    with (root / relative).open(encoding="utf-8-sig", newline="") as handle:
        rows = list(csv.DictReader(handle))
    assert rows, relative
    return rows


def check_png(root: Path, relative: str) -> None:
    data = (root / relative).read_bytes()
    assert data[:8] == b"\x89PNG\r\n\x1a\n", relative
    assert len(data) > 100, relative
    assert data[12:16] == b"IHDR", relative
    width, height = struct.unpack(">II", data[16:24])
    assert width > 0 and height > 0, relative


def check_svg(root: Path, relative: str, required: tuple[str, ...]) -> None:
    path = root / relative
    parsed = ET.parse(path)
    text = " ".join(parsed.getroot().itertext())
    for term in required:
        assert term in text, (relative, term)


def verify_phase7a(root: Path) -> dict[str, int]:
    manifest = load_json(root, PHASE7A_MANIFEST)
    assert manifest["accepted_phase"] == "7A"
    assert manifest["status"] == "ACCEPTED / FROZEN"
    assert manifest["counts"] == {
        "nodes": 20,
        "edges": 20,
        "pathway_register": 5,
        "monitoring_matrix_rows": 5,
        "evidence_crosswalk": 10,
        "uncertainty_register": 13,
    }
    return verify_artifacts(root, manifest, "Phase 7A final freeze manifest")


def verify_package(root: Path) -> dict[str, Any]:
    nodes = read_csv(root, "data/processed/networks/exposure_context_nodes.csv")
    deps = read_csv(root, EXPECTED_PACKAGE_ARTIFACTS[0])
    controls = read_csv(root, EXPECTED_PACKAGE_ARTIFACTS[1])
    evidence = read_csv(root, EXPECTED_PACKAGE_ARTIFACTS[2])
    matrix = read_csv(root, EXPECTED_PACKAGE_ARTIFACTS[3])
    assert len(deps) == 20 and len(controls) == 7 and len(evidence) == 5
    assert len(matrix) == 5 and len(matrix[0]) == 11
    assert {row["pathway_family"] for row in deps} == FAMILIES
    assert {row["pathway_family"] for row in controls} == FAMILIES
    assert {row["pathway_family"] for row in evidence} == FAMILIES
    assert {row["pathway_family"] for row in matrix} == FAMILIES
    assert all(row["source_id"] for row in controls + evidence + deps)
    node_family = {row["node_id"]: row["pathway_family"] for row in nodes}
    family_map = {
        "drinking_water_hab": "Drinking Water / HAB",
        "ambient_air": "Ambient Air",
        "legacy_contamination": "Soil / Groundwater / Legacy Contamination",
        "food_fish_recreation": "Food / Fish / Recreational Water",
        "heat": "Heat",
    }
    assert all(
        family_map[node_family[row["source_node_id"]]] == row["pathway_family"]
        and family_map[node_family[row["dependent_node_id"]]] == row["pathway_family"]
        for row in deps
    )
    assert all(row["exposure_confirmation"] == "unconfirmed" for row in evidence)
    assert all(row["dose_information"] == "unknown" for row in evidence)
    assert all(row["health_outcome_information"] == "unknown" for row in evidence)
    check_png(root, "outputs/maps/systems/24_exposure_dependencies_controls_2026.png")
    check_svg(root, "outputs/maps/systems/24_exposure_dependencies_controls_2026.svg", ("MAP 24", "Environmental condition", "Evidence gap"))
    return {"dependency_edges": 20, "controls": 7, "evidence_rows": 5, "matrix_semantic_shape": [5, 10], "matrix_stored_shape": [5, 11]}


def verify_working_manifest(root: Path) -> dict[str, int]:
    working = load_json(root, WORKING_MANIFEST)
    assert working["phase"] == "7B"
    assert working["status"] == "validated_working_package"
    assert working["counts"] == {"dependency_edges": 20, "control_register": 7, "evidence_register": 5, "matrix": [5, 10]}
    result = verify_artifacts(root, working, "Phase 7B working manifest", EXPECTED_PACKAGE_ARTIFACTS[:6])
    assert working["artifacts"]
    return result


def verify_review(root: Path) -> dict[str, bool]:
    initial = (root / REVIEW_INITIAL).read_text(encoding="utf-8")
    assert '"passed": false' in initial
    final = (root / REVIEW_FINAL).read_text(encoding="utf-8")
    match = re.search(r"```json\s*(\{.*?\})\s*```", final, flags=re.S)
    assert match, "final review JSON missing"
    verdict = json.loads(match.group(1))
    assert verdict.get("passed") is True
    for key in ("security_concerns", "logic_errors", "provenance_errors", "exposure_boundary_errors", "dependency_errors", "spatial_scale_errors", "schema_errors"):
        assert verdict.get(key) == [], key
    return {"passed": True, "blocking_arrays_empty": True, "initial_review_preserved": True, "final_review_preserved": True}


def verify_artifact_check(root: Path) -> None:
    check = load_json(root, ARTIFACT_CHECK)
    assert check == {
        "status": "passed",
        "phase": "7B",
        "dependency_edges": 20,
        "control_register": 7,
        "evidence_register": 5,
        "matrix_dimensions": [5, 11],
        "map24_valid": True,
        "phase7a_immutable": True,
        "no_exposure_score": True,
        "no_dose_or_health_model": True,
    }
    strict = load_json(root, STRICT_CHECK)
    assert strict["status"] == "passed" and strict["phase"] == "7B"
    assert strict["working_manifest_hashes_match"] is True
    assert strict["unsupported_exact_health_outcome_claims"] is False
    assert strict["unsupported_exposure_dose_illness_claims"] is False


def verify_prior_system(root: Path) -> dict[str, int]:
    expected_final_names = {Path(FINAL_MANIFEST).name, Path(SIBLING_FINAL_MANIFEST).name}
    protected: set[str] = set()
    entries = 0
    manifests = []
    for path in sorted((root / "reports").glob("phase*_freeze_manifest.json")):
        if path.name in expected_final_names:
            continue
        payload = json.loads(path.read_text(encoding="utf-8"))
        manifests.append(path.name)
        for relative, metadata in artifact_entries(payload).items():
            verify_hash(root, relative, metadata)
            protected.add(relative)
            entries += 1
    assert len(manifests) == 29, len(manifests)
    assert entries == EXPECTED_PRIOR_ENTRIES, entries
    assert len(protected) == EXPECTED_PRIOR_UNIQUE, len(protected)
    changed = set(subprocess.check_output(["git", "-C", str(root), "diff", "--name-only", SOURCE_COMMIT], text=True).splitlines())
    assert not changed.intersection(protected), sorted(changed.intersection(protected))
    return {"manifest_count": len(manifests), "entries": entries, "unique": len(protected)}


def verify_status_surfaces(root: Path) -> None:
    terms = {
        "PROJECT_STATUS.md": (
            "## phase 7b — accepted / frozen",
            "## phase 7c — accepted / frozen",
            FINAL_MANIFEST.lower(),
            SIBLING_FINAL_MANIFEST.lower(),
            "active phase: **none**",
            "phase 14 is not implemented",
            "next approved planning target",
            "great black swamp",
            "hold",
            "noncanonical",
            "intake-coordinate discrepancy",
            "unresolved",
        ),
        "docs/canon_status.md": (
            "phase 7b is **accepted / frozen**",
            "phase 7c is **accepted / frozen**",
            FINAL_MANIFEST.lower(),
            SIBLING_FINAL_MANIFEST.lower(),
            "active phase: **none**",
            "phase 14 is not implemented",
            "next approved planning target",
            "great black swamp",
            "hold",
            "noncanonical",
            "intake-coordinate discrepancy",
            "unresolved",
        ),
        "reports/current_phase_handoff.md": (
            "## phase 7b/7c final acceptance / freeze handoff",
            "phase 7b: **accepted / frozen**",
            "phase 7c: **accepted / frozen**",
            FINAL_MANIFEST.lower(),
            SIBLING_FINAL_MANIFEST.lower(),
            "active phase: **none**",
            "phase 14 is not implemented",
            "next approved planning target",
            "great black swamp",
            "hold",
            "noncanonical",
            "intake-coordinate discrepancy",
            "unresolved",
        ),
        "README.md": (
            "phase 7b is **accepted / frozen**",
            "phase 7c is **accepted / frozen**",
            FINAL_MANIFEST.lower(),
            SIBLING_FINAL_MANIFEST.lower(),
            "phase 14 is not implemented",
        ),
        "reports/README.md": (
            "phase 7b is **accepted / frozen**",
            "phase 7c is **accepted / frozen**",
            Path(FINAL_MANIFEST).name.lower(),
            Path(SIBLING_FINAL_MANIFEST).name.lower(),
        ),
        "CHANGELOG.md": (
            "sol formally accepted and froze phase 7b",
            "sol formally accepted and froze phase 7c",
            FINAL_MANIFEST.lower(),
            SIBLING_FINAL_MANIFEST.lower(),
            "active phase is none",
            "phase 14 is not implemented",
        ),
        "docs/phase_briefs/phase7b_exposure_dependencies_controls.md": (
            "status: accepted / frozen",
            FINAL_MANIFEST.lower(),
        ),
        "docs/phase_briefs/phase7c_environmental_health_futures.md": (
            "status: accepted / frozen",
            SIBLING_FINAL_MANIFEST.lower(),
        ),
    }
    for relative, required in terms.items():
        text = (root / relative).read_text(encoding="utf-8").lower()
        for term in required:
            assert term in text, (relative, term)


def verify_no_phase14(root: Path) -> None:
    for base in (root / "data/processed", root / "outputs/maps/systems", root / "src/python/systems", root / "src/R/systems"):
        for path in base.rglob("*"):
            if path.is_file():
                name = path.name.lower()
                assert "phase14" not in name and "phase_14" not in name, str(path)
                assert not name.startswith("44_"), str(path)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("root", nargs="?", default=None)
    args = parser.parse_args()
    root = Path(args.root).resolve() if args.root else Path(__file__).resolve().parents[3]
    final = load_json(root, FINAL_MANIFEST)
    assert final["accepted_phase"] == "7B"
    assert final["baseline"] == "Exposure Dependencies, Evidence Strength & Controls, 2026"
    assert final["status"] == "ACCEPTED / FROZEN"
    assert final["accepted_by"] == "Sol explicit acceptance decision supplied for this run"
    assert final["accepted_date"] == "2026-09-10"
    assert final["source_commit"] == SOURCE_COMMIT
    assert final["counts"] == {
        "dependency_edges": 20,
        "controls": 7,
        "evidence_rows": 5,
        "matrix_semantic_shape": [5, 10],
        "matrix_stored_shape": [5, 11],
    }
    assert final["phase7a_freeze_manifest"]["path"] == PHASE7A_MANIFEST
    assert final["protected_inputs"] == [PHASE7A_MANIFEST]
    assert final["phase7a_immutable"] is True
    assert final["phase14_implemented"] is False
    assert final["active_holds_preserved"] == {
        "great_black_swamp": "C — HOLD / noncanonical",
        "toledo_intake_coordinate_discrepancy": "UNRESOLVED",
    }
    assert final["deferred_maintenance_preserved"] == [
        "Phase 6B manifest status wording mismatch",
        "Phase 3A missing manifest status",
        "Phase 2A legacy worktree superseded cleanup candidate",
    ]
    final_result = verify_artifacts(root, final, "Phase 7B final freeze manifest", EXPECTED_PACKAGE_ARTIFACTS)
    working_result = verify_working_manifest(root)
    verify_artifact_check(root)
    phase7a_result = verify_phase7a(root)
    package = verify_package(root)
    prior = verify_prior_system(root)
    review = verify_review(root)
    verify_status_surfaces(root)
    verify_no_phase14(root)
    sibling = load_json(root, SIBLING_FINAL_MANIFEST)
    assert sibling["status"] == "ACCEPTED / FROZEN"
    result = {
        "status": "passed",
        "phase": "7B",
        "freeze_manifest": {"path": FINAL_MANIFEST, **final_result},
        "working_manifest": working_result,
        "package": package,
        "phase7a_integrity": phase7a_result,
        "prior_system_immutability": prior,
        "review_history": review,
        "active_holds_preserved": True,
        "deferred_maintenance_preserved": True,
        "phase14_not_implemented": True,
    }
    print(json.dumps(result, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()
